core/object/thread_runner.cpp

```diff
@@ -0,0 +1,207 @@
+	} else {
+		Callable::CallError ce;
+		Variant ret;
+		task->callable.call(nullptr, 0, ret, ce);
+	}
+
+	task->completed = true;
+	task->done_semaphore.post();
+}
+
+void ThreadRunner::_thread_function(void *p_user) {
+	ThreadData *thread = static_cast<ThreadData *>(p_user);
+	while (true) {
+		thread->scheduler->task_available_semaphore.wait();
+		if (thread->scheduler->exit_threads.load()) {
+			break;
+		}
+		thread->scheduler->_process_task();
+	}
+}
+
+ThreadRunner::TaskID ThreadRunner::add_native_task(void (*p_func)(void *), void *p_userdata, const String &p_description) {
+	task_mutex.lock();
+	// Get a free task
+	Task *task = task_allocator.alloc();
+	TaskID id = last_task++;
+	task->native_func = p_func;
+	task->native_func_userdata = p_userdata;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+ThreadRunner::TaskID ThreadRunner::add_task(const Callable &p_action, const String &p_description) {
+	task_mutex.lock();
+	// Get a free task
+	Task *task = task_allocator.alloc();
+	TaskID id = last_task++;
+	task->callable = p_action;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+bool ThreadRunner::is_task_completed(TaskID p_task_id) const {
+	task_mutex.lock();
+	const Task *const *taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
+		task_mutex.unlock();
+		ERR_FAIL_V_MSG(false, "Invalid Task ID"); // Invalid task
+	}
+
+	bool completed = (*taskp)->completed;
+	task_mutex.unlock();
+
+	return completed;
+}
+
+void ThreadRunner::wait_for_task_completion(TaskID p_task_id) {
+	task_mutex.lock();
+	Task **taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
+		task_mutex.unlock();
+		ERR_FAIL_MSG("Invalid Task ID"); // Invalid task
+	}
+	Task *task = *taskp;
+
+	if (task->waiting) {
+		String description = task->description;
+		task_mutex.unlock();
+		if (description.is_empty()) {
+			ERR_FAIL_MSG("Another thread is waiting on this task: " + itos(p_task_id)); // Invalid task
+		} else {
+			ERR_FAIL_MSG("Another thread is waiting on this task: " + description + " (" + itos(p_task_id) + ")"); // Invalid task
+		}
+	}
+
+	task->waiting = true;
+
+	task_mutex.unlock();
+
+	int *index = thread_ids.getptr(Thread::get_caller_id());
+
+	if (index) {
+		// We are an actual process thread, we must not be blocked so continue processing stuff if available.
+		while (true) {
+			if (task_available_semaphore.try_wait()) {
+				// Solve tasks while they are around.
+				_process_task();
+				continue;
+			}
+			if (task->done_semaphore.try_wait()) {
+				// If done, exit
+				break;
+			}
+			OS::get_singleton()->delay_usec(1); // Microsleep, this could be converted to waiting for multiple objects in supported platforms for a bit more performance.
```

Comment:
I think `Semaphore` is implemented based on `std::mutex` on all platforms, so it need not be platform specific. But probably it's OK as is.

Reply:
I concur, just pointing out Linux eventfd https://man7.org/linux/man-pages/man2/eventfd.2.html was created to let you block on multiple things with poll() or select().  But you really don't want to go there if you have many tasks outstanding potentially.